File: sdk/clojure/src/rockbox/transport.clj
(ns rockbox.transport
  "HTTP transport for GraphQL queries — built on `java.net.http.HttpClient`
  so the SDK has no third-party HTTP dependency.

  You normally don't call this directly; use `rockbox.core/query` or one of
  the domain APIs (`rockbox.playback`, `rockbox.library`, ...). Exposed for
  plugin authors and advanced consumers who need a stable hook."
  (:require [clojure.data.json :as json]
            [rockbox.errors :as err]
            [rockbox.util   :as util])
  (:import (java.net URI)
           (java.net.http HttpClient HttpClient$Redirect
                          HttpRequest HttpResponse$BodyHandlers
                          HttpRequest$BodyPublishers)
           (java.time Duration)))

;; ---------------------------------------------------------------------------
;; HttpClient — one cached instance per JVM (thread-safe, reusable)
;; ---------------------------------------------------------------------------

(defonce ^:private default-client
  (delay
    (-> (HttpClient/newBuilder)
        (.connectTimeout (Duration/ofSeconds 10))
        (.followRedirects HttpClient$Redirect/NORMAL)
        (.build))))

(defn- http-client ^HttpClient [client]
  (or (:http-client client) @default-client))

;; ---------------------------------------------------------------------------
;; Request building
;; ---------------------------------------------------------------------------

(defn- ->headers ^"[Ljava.lang.String;" [headers]
  (let [base ["Content-Type" "application/json"
              "Accept"       "application/json"]
        all  (into base
                   (mapcat (fn [[k v]] [(name k) (str v)]))
                   headers)]
    (into-array String all)))

(defn- build-request ^HttpRequest [{:keys [http-url timeout-ms headers]} body-json]
  (let [b (-> (HttpRequest/newBuilder)
              (.uri (URI/create http-url))
              (.timeout (Duration/ofMillis (long (or timeout-ms 15000))))
              (.POST (HttpRequest$BodyPublishers/ofString body-json)))]
    (when (seq headers)
      (.headers b (->headers headers)))
    (.header b "Content-Type" "application/json")
    (.header b "Accept"       "application/json")
    (.build b)))

(defn- send-request [client req]
  (try
    (.send (http-client client) req (HttpResponse$BodyHandlers/ofString))
    (catch java.net.ConnectException e
      (throw (err/network-error
              (str "Failed to reach Rockbox at " (:http-url client)) e)))
    (catch java.net.http.HttpConnectTimeoutException e
      (throw (err/network-error "Connect timeout" e)))
    (catch java.net.http.HttpTimeoutException e
      (throw (err/network-error "HTTP timeout" e)))
    (catch java.io.IOException e
      (throw (err/network-error (.getMessage e) e)))))

;; ---------------------------------------------------------------------------
;; Public API
;; ---------------------------------------------------------------------------

(defn execute
  "Execute a GraphQL query/mutation and return the parsed `data` map (with
  kebab-case keyword keys).

  `variables` may be a Clojure map with kebab-case keys; they are converted to
  camelCase strings before sending. Returns the kebabized `data` payload, or
  throws an `ex-info` of type `:rockbox/network` or `:rockbox/graphql`."
  ([client query]           (execute client query nil))
  ([client query variables]
   (let [vars (some-> variables util/drop-nils not-empty util/camelize-keys)
         payload (cond-> {"query" query}
                   vars (assoc "variables" vars))
         body (json/write-str payload)
         req  (build-request client body)
         resp (send-request client req)
         status (.statusCode ^java.net.http.HttpResponse resp)
         body-str (.body ^java.net.http.HttpResponse resp)]
     (when-not (<= 200 status 299)
       (throw (err/network-error (str "HTTP " status " from " (:http-url client)))))
     (let [parsed (try
                    (json/read-str body-str)
                    (catch Exception e
                      (throw (err/network-error
                              (str "Invalid JSON response: " (.getMessage e)) e))))
           errors (get parsed "errors")]
       (when (and errors (seq errors))
         (throw (err/graphql-error (mapv util/kebabize-keys errors))))
       (-> (get parsed "data") util/kebabize-keys)))))

(defn execute-field
  "Convenience wrapper: execute the query, then pluck a single top-level
  field from the response. Most domain APIs use this for one-line bodies."
  ([client query field]
   (get (execute client query) field))
  ([client query variables field]
   (get (execute client query variables) field)))

File: sdk/elixir/lib/rockbox/transport.ex
defmodule Rockbox.Transport do
  @moduledoc """
  Low-level GraphQL HTTP transport. You normally don't call this directly —
  the `Rockbox.*` API modules wrap it. Exposed because `Rockbox.query/3` and
  the plugin system need a stable hook for raw queries.
  """

  alias Rockbox.{Client, Error, GraphQLError, NetworkError, Util}

  @type variables :: map() | Keyword.t() | nil

  @doc """
  Execute a GraphQL operation and return `{:ok, data}` or `{:error, exception}`.

  `data` is returned as-is (a map), not coerced to a struct — the API modules
  handle struct conversion. Variables are passed through `Rockbox.Util.camelize/1`
  to convert snake_case keys to the camelCase the server expects.
  """
  @spec execute(Client.t(), String.t(), variables()) ::
          {:ok, map()} | {:error, Exception.t()}
  def execute(%Client{} = client, query, variables \\ nil) do
    body = %{query: query, variables: encode_variables(variables)}

    case do_post(client, body) do
      {:ok, %{status: status, body: body}} when status in 200..299 ->
        case body do
          %{"errors" => [_ | _] = errors} ->
            {:error, GraphQLError.exception(Util.atomize(errors))}

          %{"data" => data} ->
            {:ok, data}

          other ->
            {:error, %Error{message: "Unexpected GraphQL response shape: #{inspect(other)}"}}
        end

      {:ok, %{status: status, body: body}} ->
        {:error,
         %NetworkError{
           message: "HTTP #{status} from #{client.http_url}",
           cause: body
         }}

      {:error, reason} ->
        {:error,
         %NetworkError{
           message: "Failed to reach rockboxd at #{client.http_url}",
           cause: reason
         }}
    end
  end

  @doc "Same as `execute/3` but raises on error."
  @spec execute!(Client.t(), String.t(), variables()) :: map()
  def execute!(client, query, variables \\ nil) do
    case execute(client, query, variables) do
      {:ok, data} -> data
      {:error, exception} -> raise exception
    end
  end

  # ---------------------------------------------------------------------------
  # Internal
  # ---------------------------------------------------------------------------

  defp do_post(%Client{} = client, body) do
    headers = [
      {"content-type", "application/json"},
      {"accept", "application/json"} | client.headers
    ]

    Req.post(
      client.http_url,
      json: body,
      headers: headers,
      receive_timeout: client.timeout,
      retry: false
    )
  end

  defp encode_variables(nil), do: nil
  defp encode_variables(vars) when map_size(vars) == 0, do: nil
  defp encode_variables(vars), do: Util.camelize(vars)
end
